src/peft/tuners/hira/layer.py

```diff
@@ -0,0 +1,937 @@
+            hira_dropout_layer = nn.Dropout(p=hira_dropout)
+        else:
+            hira_dropout_layer = nn.Identity()
+
+        self.hira_dropout.update(nn.ModuleDict({adapter_name: hira_dropout_layer}))
+        # Actual trainable parameters
+        self.hira_A.update(nn.ParameterDict({adapter_name: nn.Parameter(torch.randn(r, self.in_features))}))
+        self.hira_B.update(nn.ParameterDict({adapter_name: nn.Parameter(torch.randn(self.out_features, r))}))
+
+        # for inits that require access to the base weight, use gather_param_ctx so that the weight is gathered when using DeepSpeed
+        if init_hira_weights:
+            self.reset_hira_parameters(adapter_name, init_hira_weights)
+        self._move_adapter_to_device_of_base_layer(adapter_name)
+
+        self.set_adapter(self.active_adapters)
+
+    def reset_hira_parameters(self, adapter_name, init_hira_weights):
+        if init_hira_weights is False:
+            return
+
+        if adapter_name in self.hira_A.keys():
+            if init_hira_weights is True:
+                # initialize A the same way as the default for nn.Linear and B to zero
+                nn.init.kaiming_uniform_(self.hira_A[adapter_name], a=math.sqrt(5))
+            elif init_hira_weights.lower() == "gaussian":
+                nn.init.normal_(self.hira_A[adapter_name], std=1 / self.r[adapter_name])
+            else:
+                raise ValueError(f"Unknown initialization {init_hira_weights=}")
+            nn.init.zeros_(self.hira_B[adapter_name])
+        if adapter_name in self.hira_embedding_A.keys():
+            # Initialize A to zeros and B the same way as the default for nn.Embedding, see:
+            nn.init.zeros_(self.hira_embedding_A[adapter_name])
+            nn.init.normal_(self.hira_embedding_B[adapter_name])
+
+    def _cache_store(self, key: str, value: Any) -> None:
+        self._caches[key] = value
+
+    def _cache_pop(self, key: str) -> Any:
+        value = self._caches.pop(key)
+        return value
+
+    def _check_forward_args(self, x, *args, **kwargs):
+        """Check if the arguments are compatible with the configs and state of the model"""
+        adapter_names = kwargs.get("adapter_names", None)
+        if adapter_names is None:
+            return
+
+        if len(x) != len(adapter_names):
+            msg = (
+                "Length of `adapter_names` should be the same as the number of inputs, but got "
+                f"{len(adapter_names)} and {len(x)} respectively."
+            )
+            raise ValueError(msg)
+
+        if self.merged:
+            # It is unclear what would be the right thing to do if users pass adapter_names and there are merged
+            # adapters. Therefore, it is better to raise an error in this case.
+            msg = "Cannot pass `adapter_names` when there are merged adapters, please call `unmerge_adapter` first."
+            raise ValueError(msg)
+
+    def _mixed_batch_forward(
```

Comment:
This method and its calls below can be removed if we drop `_enable_peft_forward_hooks`.